pkg/controller/direct/compute/firewallpolicyrule/firewallpolicyrule_controller.go

```diff
@@ -0,0 +1,290 @@
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	firewallPolicyRule.Priority = nil
+
+	op := &gcp.Operation{}
+	updated := &computepb.FirewallPolicyRule{}
+
+	updateReq := &computepb.PatchRuleFirewallPolicyRequest{
+		FirewallPolicyRuleResource: firewallPolicyRule,
+		FirewallPolicy:             a.firewallPolicy,
+		Priority:                   direct.PtrTo(int32(a.priority)),
+	}
+	op, err = a.firewallPoliciesClient.PatchRule(ctx, updateReq)
+	if err != nil {
+		return fmt.Errorf("updating ComputeFirewallPolicyRule %d: %w", a.priority, err)
+	}
+	if !op.Done() {
+		err = op.Wait(ctx)
+		if err != nil {
+			return fmt.Errorf("waiting ComputeFirewallPolicyRule %d update failed: %w", a.priority, err)
+		}
+	}
+	log.V(2).Info("successfully updated ComputeFirewallPolicyRule", "priority", a.priority)
+
+	// Get the updated resource
+	updated, err = a.get(ctx)
+	if err != nil {
+		return fmt.Errorf("getting ComputeFirewallPolicyRule %d: %w", a.priority, err)
+	}
+
+	status := &krm.ComputeFirewallPolicyRuleStatus{
+		RuleTupleCount: direct.PtrTo(int64(*updated.RuleTupleCount)),
+		Kind:           direct.PtrTo("compute#firewallPolicyRule"),
+	}
+	return updateOp.UpdateStatus(ctx, status, nil)
+}
+
+func (a *firewallPolicyRuleAdapter) Export(ctx context.Context) (*unstructured.Unstructured, error) {
+	if a.actual == nil {
+		return nil, fmt.Errorf("firewallPolicyRule %d not found", a.priority)
+	}
+
+	mc := &direct.MapContext{}
+	spec := ComputeFirewallPolicyRuleSpec_FromProto(mc, a.actual)
+	specObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(spec)
+	if err != nil {
+		return nil, fmt.Errorf("error converting firewallPolicyRule spec to unstructured: %w", err)
+	}
+
+	u := &unstructured.Unstructured{
+		Object: make(map[string]interface{}),
+	}
+	u.SetGroupVersionKind(krm.ComputeFirewallPolicyRuleGVK)
+
+	if err := unstructured.SetNestedField(u.Object, specObj, "spec"); err != nil {
+		return nil, fmt.Errorf("setting spec: %w", err)
+	}
+
+	return u, nil
+}
+
+// Delete implements the Adapter interface.
+func (a *firewallPolicyRuleAdapter) Delete(ctx context.Context, deleteOp *directbase.DeleteOperation) (bool, error) {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("deleting ComputeFirewallPolicyRule", "priority", a.priority)
+
+	delReq := &computepb.RemoveRuleFirewallPolicyRequest{
+		FirewallPolicy: a.firewallPolicy,
+		Priority:       direct.PtrTo(int32(a.priority)),
+	}
+	op, err := a.firewallPoliciesClient.RemoveRule(ctx, delReq)
+
+	if err != nil {
+		return false, fmt.Errorf("deleting ComputeFirewallPolicyRule %d: %w", a.priority, err)
+	}
+	if !op.Done() {
+		err = op.Wait(ctx)
+		if err != nil {
+			return false, fmt.Errorf("waiting ComputeFirewallPolicyRule %d delete failed: %w", a.priority, err)
+		}
+	}
+	log.V(2).Info("successfully deleted ComputeFirewallPolicyRule", "priority", a.priority)
+
+	// Get the deleted rules
+	_, err = a.get(ctx)
+	if err != nil {
+		return true, nil
+	}
```

Comment:
We don't need this. It makes an additional GCP call and is covered by above code ln268-274

Reply:
I added this because there was no additional GET call to verify the rule has been deleted. Removing the GET call here will cause a missing GET request compared to the TF controller.

Reply:
I think it's OK to have one fewer get, if TF is doing it wrong :-)

If you want to keep it to make the golden output look the same (until we fully get rid of the TF) then consider something like this:

```
// Temporarily add an extra GET to match Terraform behaviour
a.get(ctx)
```